### ISSUE TITLE
Snapping Control

### ISSUE THREAD

**Issue by bjornharrtell** [openlayers/ol3#1164]

**Comment by fperucic** [comment]
I need snap functionalty for a project I'm on, so I'm currenty working on solving this problem.
My idea is to mI need snap functionalty for a project I'm on, so I'm currenty working on solving this problem.
My general idea is to extract snap functionalty from ol.interaction.Modify and make a `ol.Snap` object which receives some parameters:

* features: `ol.Collection.<ol.Feature>` 
	* features to snap to
* interactions: `Array.<ol.interaction.Pointer>`
	* listen for events on interactions (Modify, Draw or custom interaction)
* pixelTolerance

Each interaction sholud expose it's current coordinate via snap event. And if `ol.Snap` is listening for `SNAPME` events on the interaction it whould bind the snapped coordinates to the event. The interaction whould then use the snapped coordinate.

```javascript
var event = new ol.SnapEvent(
	ol.SnapEventType.SNAPME,
	event.coordinates
);
this.dispatchEvent(event);
var possiblySnappedCoordinates = event.snappedCoordinate;
/* event.snappedCoordinate is used as current coordinate insead of event.coordinate */
```

Place to fire a *SNAPME* event in `interaction.Draw` is `ol.interaction.Draw.prototype.handlePointerMove_`,
and in `interaction.Modify` is `ol.interaction.Modify.handleDragEvent_`.


This is my idea of solving this problem, any feedback is appreciated.

**Comment by ttsiodras** [comment]
@lenoil Yes, isVoid just checks for undefined and null, and no, I wrote the code in TypeScript - but it's very easily translatable, the TypeScript concepts are pretty straightforward to map. If you have any problems with some parts of the syntax, ask away.

**Comment by ahocevar** [comment]
Fixed with #3109.
